### ISSUE TITLE
Receiving 'We are completely uploaded and fine' but also receive 400 Bad Request when uploading encrypted files

### ISSUE THREAD

**Issue by abraunegg** [OneDrive/onedrive-api-docs#443]
When uploading encrypted files (eCryptFS) I am getting what appears to be dual responses from OneDrive:

```
Uploading new file: ./564df3d9-26c9-bc34-6e42-a042e0948d28/encrypted/ECRYPTFS_FNEK_ENCRYPTED.FWZT5friQSJghUbeFfg6iiUWjaAjlvB-fJCt0vHWaCWh6hoR8qe7Lod5QU--/ECRYPTFS_FNEK_ENCRYPTED.FWZT5friQSJghUbeFfg6iiUWjaAjlvB-fJCtcyXqjdW62Hyj6K4CoHTcVk--/ECRYPTFS_FNEK_ENCRYPTED.FWZT5friQSJghUbeFfg6iiUWjaAjlvB-fJCt5cZkg2tDk-hG1Yomr3GZNU--/ECRYPTFS_FNEK_ENCRYPTED.FWZT5friQSJghUbeFfg6iiUWjaAjlvB-fJCt7cgYx6giUdlILllImnH8tE--/ECRYPTFS_FNEK_ENCRYPTED.FXZT5friQSJghUbeFfg6iiUWjaAjlvB-fJCtJ8Iq-w6wMVgPsgiid8qhnEUCyRy1aAum0VvjSpqufP--
* Found bundle for host api.onedrive.com: 0x10005e0
* Re-using existing connection! (#1) with host api.onedrive.com
* Connected to api.onedrive.com (204.79.197.213) port 443 (#1)
> PUT /v1.0/drive/root:/.%2F564df3d9-26c9-bc34-6e42-a042e0948d28%2Fencrypted%2FECRYPTFS_FNEK_ENCRYPTED.FWZT5friQSJghUbeFfg6iiUWjaAjlvB-fJCt0vHWaCWh6hoR8qe7Lod5QU--%2FECRYPTFS_FNEK_ENCRYPTED.FWZT5friQSJghUbeFfg6iiUWjaAjlvB-fJCtcyXqjdW62Hyj6K4CoHTcVk--%2FECRYPTFS_FNEK_ENCRYPTED.FWZT5friQSJghUbeFfg6iiUWjaAjlvB-fJCt5cZkg2tDk-hG1Yomr3GZNU--%2FECRYPTFS_FNEK_ENCRYPTED.FWZT5friQSJghUbeFfg6iiUWjaAjlvB-fJCt7cgYx6giUdlILllImnH8tE--%2FECRYPTFS_FNEK_ENCRYPTED.FXZT5friQSJghUbeFfg6iiUWjaAjlvB-fJCtJ8Iq-w6wMVgPsgiid8qhnEUCyRy1aAum0VvjSpqufP--:/content?@name.conflictBehavior=fail HTTP/1.1
User-Agent: Phobos-std.net.curl/2.071 (libcurl/7.29.0)
Host: api.onedrive.com
Accept: */*
Content-Type: application/octet-stream
Authorization: bearer <REDACTED>
Content-Type: application/octet-stream
Content-Length: 12288
Expect: 100-continue

< HTTP/1.1 100 Continue
* We are completely uploaded and fine
< HTTP/1.1 400 Bad Request
< Via: 1.1 BY3302____PAP248 (wls-colorado)
< Content-Length: 113
< Content-Type: application/json
< Server: Microsoft-IIS/8.5
< P3P: CP="BUS CUR CONo FIN IVDo ONL OUR PHY SAMo TELo"
< X-WLSPROXY: BY3302____PAP248
< X-MSNSERVER: SN3302____PAP054
< Strict-Transport-Security: max-age=31536000; includeSubDomains
< X-QosStats: {"ApiId":0,"ResultType":2,"SourcePropertyId":0,"TargetPropertyId":42}
< X-ThrowSite: 475f.c096
< X-AsmVersion: UNKNOWN; 19.49.0.0
< X-MSEdge-Ref: Ref A: 2BF1D0A8A5C44A629EEC4AE3287FBBF0 Ref B: 9B779DD714CC91855522B0D2ED5FEF01 Ref C: Sat Sep 10 00:07:02 2016 PST
< Date: Sat, 10 Sep 2016 07:07:01 GMT
< 
* Connection #1 to host api.onedrive.com left intact
OneDrive returned a 'HTTP 400 - Bad Request' - gracefully handling error
No 'id' key found in JSON array - skipping updating DB (uploadNewFile)
```

Looking at OneDrive, the folder structure exists, however the file certainly did not get uploaded.

Other folders / files appear to be getting uploaded, in other paths and depths as I was initially thinking it was a path length issue.

Any suggestions as to what could be going on here?

Happy to provide an archive of the encrypted files if required.

Best regards,

Alex

**Comment by ificator** [comment]
AFC (away from code), but at a glance it may be the system getting confused with the . at the beginning of the path you're providing. I can confirm on Monday.

**Comment by abraunegg** [comment]
Thanks for looking into this

Also using the same client to upload, same file set but not encrypted, the files all upload without issue, with the . at the beginning of the file set.

Best regards,

Alex

**Comment by ificator** [comment]
So the failure here is actually a `pathIsTooLong` resulting from the size of the encrypted path.

Unfortunately, for legacy reasons, the limits are different between OneDrive Consumer and OneDrive Business and so the API needs to recommends the smaller of the two. This happens to be OneDrive Business with its 256 character limit - see [this KB article](https://support.microsoft.com/en-us/kb/3125202). If you only care about supporting OneDrive consumer then you can target a maximum length of around 430 - see [this user voice response](https://onedrive.uservoice.com/forums/262982-onedrive/suggestions/6325774-enable-long-file-paths-aka-file-name-too-long-p).

**Comment by abraunegg** [comment]
Many thanks for looking into this. The Linux OneDrive client (https://github.com/skilion/onedrive) only supports OneDrive Consumer at the moment, so what I will have to do is update my branch to validate that the length of the string is < 430 characters for the file that is being uploaded. I will also have to update MyNAS Storage Appliance (www.mynas.com.au) to articulate that storing encrypted content is not supported (due to the potential long paths which are created) at this point in time.

Is there any way to get OneDrive to return a new 4xx error articulating path too long rather than a generic 400 Bad Request?

**Comment by ificator** [comment]
When we defined this particular scenario we looked over the status codes and nothing aligned. `414 URI Too Long` was close, but it implied something other than the actual problem. In those cases we generally fell back to the more generic `400 Bad Request` but were careful to make sure that the response itself included details for clients that want them. In this case the response would have looked like:

```JSON
{
    "error": {
        "code": "invalidRequest",
        "message": "Path exceeds maximum length",
        "innererror": {
            "code": "pathIsTooLong",
        }
    }
}
```

For more information on detailed error codes have a read of the documentation just below [this table](https://dev.onedrive.com/misc/errors.htm#the-code-property).
